### DISCUSSION TITLE
Question: How to avoid pickling multiple times the same python object if "send" to multiple ranks

### DISCUSSION THREAD

**Discussion by Deathn0t** [mpi4py/mpi4py#182] [Q&A] [answered]
Hello,

I am looking for suggestions/pointers for a tiny optimisation that I would like to test.  Let's assume I have the following example:

```python
o = {"a": 0, "b": 1}
for i in range(comm.Get_size()):
  comm.send(o, dest=i)
```

How could I avoid pickling multiple times the `o` object in this case? I want to do this for `isend`.

**Comment by dalcinl** [comment, marked as answer]
Well, for that exact code you used, you just don't use `send()`, use `bcast()`. It would be nice to have lowercase `ibcast()`, but it would be hard to implement without quite a bit of changes and logic to the `Request` class.

But If you really insist with using point-to-point and optimizing pickle away... Just call `data = MPI.pickle.dumps(obj)` yourself and use the uppercase method `req = comm.Isend(data, ...)` on the sending side. On the receiving side, just use the usual lowercase method `obj = comm.recv(...)`.

If the object you are trying to communicate is large and buffer-like with support for out-of-band pickle with protocol version 5 (for example, a large NumPy array), then you should probably look at [`mpi4py.util.plk5`](https://mpi4py.readthedocs.io/en/stable/mpi4py.util.pkl5.html).

**Reply by Deathn0t** [reply]
Thanks for the detailed replied that's perfect!

To give you a bit more details, I want my ranks to communicate asynchronously with each other so I have each rank to:

```python
while not done:
  0. generate `data` (takes some time which may vary)
  1. send `data` to all other ranks
  2. collect all received data
``` 

If I have just two ranks it is possible that this little logic is executed twice by rank 0 before rank 1 calls it just once.

Initially I was looking for `ibcast` but decided to implement with `isend/irecv` instead.

The two implemented functions I used are here
1. `send_all`: [GitHub link](https://github.com/deephyper/deephyper/blob/develop/deephyper/search/hps/_dmbs_mpi.py#L206)
2. `recv_any`: [Github link](https://github.com/deephyper/deephyper/blob/develop/deephyper/search/hps/_dmbs_mpi.py#L235)

I already managed to have this running on 4k ranks but I noticed when moving from 2k ranks to 4k ranks that the time is doubling (which is normal given the loop).

**Reply by dalcinl** [reply]
This [line](https://github.com/deephyper/deephyper/blob/develop/deephyper/search/hps/_dmbs_mpi.py#L264) is a bit suspicious. If there is a pickle error, than can only happen after the `MPI_Request` handle has been tested and set to `MPI_REQUEST_NULL`, so you should not call `cancel()` on the request object. Does it make sense? Maybe I'm missing something...

**Reply by Deathn0t** [reply]
Okay. I thought maybe in case of error the memory used by the request was not freed?

**Reply by dalcinl** [reply]
The MPI request handle is already de allocated. Python’s garbage collector should take care of the rest.

**Reply by Deathn0t** [reply]
Good to know thanks!

**Reply by Deathn0t** [reply]
When scaling this to 8192 ranks I end up with a "segmentation fault" do you know what could be the cause of this?

**Reply by dalcinl** [reply]
No idea! Are you using Open MPI?

**Reply by Deathn0t** [reply]
No I am using `mpich` library provided on the system I use: [MPI-Theta](https://www.alcf.anl.gov/support-center/theta/mpi-theta) and I built `mpi4py` against it with the following

```console
cd mpi4py/
sed -i "s/## mpicc .*/mpicc = cc/g" mpi.cfg
sed -i "s/## mpicxx .*/mpicxx = CC/g" mpi.cfg
CC=cc CXX=CC python setup.py build
CC=cc CXX=CC python setup.py install
cd ..
```

**Reply by dalcinl** [reply]
I'm clueless. I'm afraid you will have to figure out the source of the segfault, and I don't think it is mpi4py's fault.

FYI, you can just do:

```
MPICC=cc python setup.py install
```

or perhaps better:

```
MPICC=cc python -m pip install .
```

**Reply by Deathn0t** [reply]
I tried! Thanks for the reply :)
